tools/codegen/api/dispatcher.py

```diff
@@ -75,74 +74,86 @@ def arguments(func: FunctionSchema) -> Sequence[DispatcherArgument]:
             for la in legacy_dispatcher.arguments(func)
         ]
 
-# TODO GATHER is only needed for non-c10-full ops, remove later.
-ProcessTensoroptions = Enum('ProcessTensoroptions', ('GATHER', 'SCATTER', 'PASS_THROUGH'))
-
-
 # Given a set of CppArguments in scope, return a sequence of dispatcher
 # expressions that translate the cpp API into dispatcher API
-def cppargument_exprs(a: CppArgument,
-                      *,
-                      tensor_options: Optional[CppArgument],
-                      process_tensoroptions: ProcessTensoroptions = ProcessTensoroptions.PASS_THROUGH
-                      ) -> Sequence[DispatcherExpr]:
-    if isinstance(a.argument, TensorOptionsArguments):
-        if process_tensoroptions == ProcessTensoroptions.SCATTER:
-            ta = a.argument
+#
+# WARNING: This is unsound if you pass it CppArgument when you were
+# supposed to pass it CppTensorOptionsArguments, it will directly
+# translate device to device, which will give you the wrong signature
+# for dispatcher.  If Argument "knew" that it was part of a
+# TensorOptions that would help us dynamically test for this case
+def cppargument_exprs(
+    a: CppArgumentPack,
+    *, tensor_options: Optional[CppArgument]
+) -> Sequence[DispatcherExpr]:
+    if isinstance(a, CppSingleArgumentPack):
```

Comment:
I think you've bought some dynamic complexity at this level by factoring the transformation from `CppSignature` to `Sequence[DispatcherExpr]` through a `Sequence[CppArgumentPack]`. If the builder had access to information `CppSignature` had (or could have) about the signature as a whole, it could make direct calls to smaller functions defined over `CppArgumentPack`s subtypes.

I don't know without trying it whether this would be a win, but I mention it bc it feels like the current setup might be falling prey to that thing that happens when trans-sequence structural info gets pushed down into sequence elements, and map/flatMap element handlers wind up needing to reconstruct chunks of the larger structural situation from context.

Reply:
I think https://github.com/pytorch/pytorch/pull/45890#discussion_r503655363 addresses this

Reply:
See there - you can definitely avoid the antipattern I'm mentioning here without sacrificing the expressiveness you need to track self and tensor options that show up in different places. But I wouldn't suggest retroactively tearing up the sidewalk at this point.